### DISCUSSION TITLE
How to work with Build Errors in Helix

### DISCUSSION THREAD

**Discussion by daveb68** [helix-editor/helix#5541] [Q&A]
How do you run and work with the output of build tools in the helix editor?

For example, "make" on a simple c project create the following lines between stdout and stderr:

```
cc    -c -o main.o main.c
main.c: In function ‘main’:
**main.c:66:4: error: unknown type name ‘this’**
   66 |    this
      |    ^~~~
make: *** [<builtin>: main.o] Error 1
```

Every code editor I have ever used allows starting a build and stepping through the build errors.    Vim does it with its quickfix list.  IDE's tend to do it with an errors pane.   

Is Helix currently limited to using the diagnostics kicked out by the LSP protocol for fixing compile/linting errors?  Or am I missing something?

Is there any quickfix support or ability to load the output of a build tool into a new buffer and jump to file positions from that file, such as the 'F' key in vim?

**Comment by quantonganh** [comment]
If you are using [WezTerm](https://github.com/wez/wezterm), try this:

- create a script to run your code, for e.g:

`wezterm-split-pane.sh`:

```sh
#!/bin/sh

pane_id=$(wezterm cli get-pane-direction down)
if [ -z "${pane_id}" ]; then
  pane_id=$(wezterm cli split-pane)
fi

wezterm cli activate-pane-direction --pane-id $pane_id down
```

`run.sh`:

```sh
#!/bin/sh

source wezterm-split-pane.sh

program=$(wezterm cli list | awk -v pane_id="$pane_id" '$3==pane_id { print $6 }')
if [ "$program" = "lazygit" ]; then
  echo "q" | wezterm cli send-text --pane-id $pane_id --no-paste
fi

filename="$1"
basedir=$(dirname "$filename")
basename=$(basename "$filename")
basename_without_extension="${basename%.*}"
extension="${filename##*.}"

case "$extension" in
  "c")
    run_command="clang -lcmocka -lmpfr -Wall -O3 $filename -o $basedir/$basename_without_extension && $basedir/$basename_without_extension"
    ;;
  "go")
    run_command="go run $filename"
    ;;
  "scm"|"rkt")
    run_command="racket $filename"
    ;;
esac

echo "${run_command}" | wezterm cli send-text --pane-id $pane_id --no-paste
# wezterm cli activate-pane-direction up
```

- add the following into your Helix config file (need [this PR](https://github.com/helix-editor/helix/pull/6979) to expand the filename):

```toml
[keys.normal.";"]
r = ":sh run.sh %val{filename} > /tmp/run.log 2>&1"
```

Now you can just press `;`, followed by `r` to run your code.

And to jump to the error positions, you need:
- turn your local file path into a [hyperlink](https://wezfurlong.org/wezterm/config/lua/config/hyperlink_rules.html):

```lua
config.hyperlink_rules = wezterm.default_hyperlink_rules()

table.insert(config.hyperlink_rules, {
  regex = '^/[^/\r\n]+(?:/[^/\r\n]+)*:\\d+:\\d+',
  format = "$0",
})
```

- [open](https://wezfurlong.org/wezterm/config/lua/window-events/open-uri.html) that file in the Helix pane:

```lua
wezterm.on("open-uri", function(window, pane, uri)
  local user = os.getenv("USER")
  local start, _ = string.find(uri, '/Users/' .. user)
  if start == 1 then
    local hx_pane = pane:tab():get_pane_direction("Up")
    local action = wezterm.action.SendString(":open " .. uri .. "\r\n")

    -- and spawn it!
    window:perform_action(action, hx_pane);
    -- prevent the default action from opening in a browser
    return false
  end
  -- otherwise, by not specifying a return value, we allow later
  -- handlers and ultimately the default action to caused the
  -- URI to be opened in the browser
end)
```

Hope this helps.

**Reply by nuke-web3** [reply]
Just found your great blog that speaks to this :bow: https://quantonganh.com/2023/07/21/jump-to-build-error-helix as part of this _fantastic series_ https://quantonganh.com/2023/08/19/turn-helix-into-ide.md :star2: 

I am trying to get this kinda functionality working - the `cargo....` in VSCode term -> jump to line is _so powerful_ for me. But I use [Alacritty](https://alacritty.org/) as my term for now... I am not sure I wanna jump to another just for that kinda feature... you know any way one could do what WezTerm is doing for other terms? 

Edit: this is cool enough to make the jump to Wezterm _just_ to try this out. Thanks for the motivation :grinning:

**Comment by nuke-web3** [comment]
While specifically for rust, this is an exciting Neovim focused project to make errors better in UX and also jump to the line of the error in Neovim... other editor are possible: https://github.com/alopatindev/cargo-limit#text-editoride-integrations :eyes:
